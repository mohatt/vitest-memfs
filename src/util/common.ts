import { vi, expect } from 'vitest'

/**
 * Imports the real `fs/promises` module, bypassing Vitest mocks.
 *
 * A direct import works in most cases, but if `fs` is mocked
 * in a Vitest setup file it will return the mock instead.
 */
export async function importActualFS() {
  return vi.importActual<typeof import('fs/promises')>('fs/promises')
}

export function isPlainObject(value: unknown): value is Record<string, any> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

type MatchersObject = Parameters<(typeof expect)['extend']>[0]

/**
 * Creates and returns a matcher function.
 * We could wrap the matcher function in the future to warn on un-awaited promises.
 */
export function createMatcher<T extends keyof MatchersObject>(matcher: T, fn: MatchersObject[T]) {
  Object.defineProperty(fn, 'name', { value: matcher })
  return fn
}
