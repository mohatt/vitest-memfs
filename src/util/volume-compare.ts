import { createHash } from 'node:crypto'
import { isText } from 'istextorbinary'
import type { VolumeMap, VolumeMapEntry } from './volume.js'

export type VolumeCompareListMatch = 'exact' | 'ignore-extra' | 'ignore-missing'
export type VolumeCompareReportType = 'first' | 'all'

export interface VolumeCompareOptions {
  listMatch?: VolumeCompareListMatch
  report?: VolumeCompareReportType
}

export interface VolumeCompareResult {
  pass: boolean
  message: () => string
  actual?: unknown
  expected?: unknown
}

/**
 * Compare two volume maps.
 *
 * @param received map from volume under test
 * @param expected map from reference (snapshot or other volume)
 * @param options extra options
 */
export function compareVolumeMaps(
  received: VolumeMap,
  expected: VolumeMap,
  options?: VolumeCompareOptions,
): VolumeCompareResult {
  if (options?.report === 'all') {
    return compareVolumeMapsFull(received, expected, options)
  }

  const listMatch = options?.listMatch
  // make sorted arrays for error reporting and better diffing
  const actualFiles = Object.keys(received).sort()
  const expectedFiles = Object.keys(expected).sort()

  let listMismatchResult: {
    reason: string
    actual: string[]
    expected: string[]
  } | null = null

  switch (listMatch) {
    case 'ignore-extra': {
      const missing = expectedFiles.filter((f) => !(f in received))
      if (missing.length > 0) {
        listMismatchResult = {
          reason: `volume is missing ${missing.length} expected file${missing.length > 1 ? 's' : ''}`,
          actual: actualFiles.filter((f) => f in expected),
          expected: expectedFiles,
        }
      }
      break
    }

    case 'ignore-missing': {
      const extra = actualFiles.filter((f) => !(f in expected))
      if (extra.length > 0) {
        listMismatchResult = {
          reason: `volume has ${extra.length} unexpected file${extra.length > 1 ? 's' : ''}`,
          actual: actualFiles,
          expected: expectedFiles.filter((f) => f in received),
        }
      }
      break
    }

    case 'exact':
    default: {
      if (
        actualFiles.length !== expectedFiles.length ||
        !actualFiles.every((f, i) => f === expectedFiles[i])
      ) {
        listMismatchResult = {
          reason: 'directory structure didnâ€™t match',
          actual: actualFiles,
          expected: expectedFiles,
        }
      }
    }
  }

  if (listMismatchResult) {
    return {
      pass: false,
      message: () => listMismatchResult.reason,
      actual: listMismatchResult.actual,
      expected: listMismatchResult.expected,
    }
  }

  // compare contents
  const filesToCheck = listMatch === 'ignore-missing' ? actualFiles : expectedFiles
  for (const file of filesToCheck) {
    const exp = expected[file]
    const act = received[file]
    if (exp.type !== act.type) {
      return {
        pass: false,
        message: () => `path type mismatch at \`${file}\``,
        actual: makeEntryPreview(file, act),
        expected: makeEntryPreview(file, exp),
      }
    }
    if (exp.type === 'file') {
      const expBuff = exp.data
      const actBuff = (act as typeof exp).data
      if (!expBuff.equals(actBuff)) {
        return {
          pass: false,
          message: () => `mismatch in file \`${file}\``,
          actual: makeEntryPreview(file, act),
          expected: makeEntryPreview(file, exp),
        }
      }
    } else if (exp.type === 'symlink') {
      const expTarget = exp.target
      const actTarget = (act as typeof exp).target

      if (expTarget !== actTarget) {
        return {
          pass: false,
          message: () => `symlink target mismatch at \`${file}\``,
          actual: makeEntryPreview(file, act),
          expected: makeEntryPreview(file, exp),
        }
      }
    }
  }

  // everything is matched at this point
  return {
    pass: true,
    message: () => 'volumes matched',
  }
}

export function compareVolumeMapsFull(
  received: VolumeMap,
  expected: VolumeMap,
  options?: Omit<VolumeCompareOptions, 'report'>,
) {
  const listMatch = options?.listMatch
  const actualDiff: Record<string, unknown> = {}
  const expectedDiff: Record<string, unknown> = {}
  let missingCount = 0
  let extraCount = 0
  let diffCount = 0

  function addDiff(f: string, exp?: VolumeMapEntry, act?: VolumeMapEntry) {
    if (act) actualDiff[f] = makeEntryPreview(f, act)
    if (exp) expectedDiff[f] = makeEntryPreview(f, exp)
  }

  function addMatch(f: string) {
    actualDiff[f] = {}
    expectedDiff[f] = {}
  }

  const filesToCheck =
    listMatch === 'ignore-extra'
      ? expected
      : listMatch === 'ignore-missing'
        ? received
        : { ...received, ...expected }
  for (const f in filesToCheck) {
    const exp = expected[f]
    const act = received[f]

    if (exp && act) {
      if (exp.type !== act.type) {
        addDiff(f, exp, act)
        diffCount++
      } else if (exp.type === 'file') {
        if (!exp.data.equals((act as typeof exp).data)) {
          addDiff(f, exp, act)
          diffCount++
        } else {
          addMatch(f)
        }
      } else if (exp.type === 'symlink') {
        if (exp.target !== (act as typeof exp).target) {
          addDiff(f, exp, act)
          diffCount++
        } else {
          addMatch(f)
        }
      } else {
        addMatch(f)
      }
    } else if (exp && !act && listMatch !== 'ignore-missing') {
      addDiff(f, exp)
      missingCount++
    } else if (act && !exp && listMatch !== 'ignore-extra') {
      addDiff(f, null, act)
      extraCount++
    }
  }

  const total = missingCount + extraCount + diffCount
  if (total > 0) {
    const parts: string[] = []
    if (missingCount) parts.push(`${missingCount} missing path${missingCount > 1 ? 's' : ''}`)
    if (extraCount) parts.push(`${extraCount} unexpected path${extraCount > 1 ? 's' : ''}`)
    if (diffCount) parts.push(`${diffCount} mismatched content`)

    return {
      pass: false,
      message: () =>
        parts.length === 1
          ? `Found ${parts[0]}` //
          : `Found ${total} mismatches: ${parts.join(', ')}`,
      actual: actualDiff,
      expected: expectedDiff,
    }
  }

  return { pass: true, message: () => 'volumes matched' }
}

function makeEntryPreview(name: string, entry: VolumeMapEntry) {
  if (entry.type === 'empty-dir') {
    return new Directory()
  }
  if (entry.type === 'file') {
    if (isText(name, entry.data)) {
      return new TextFile(entry.data)
    }
    return new BinaryFile(entry.data)
  }
  return new Symlink(entry.target)
}

class TextFile {
  data: string
  constructor(buff: Buffer) {
    this.data = buff.toString('utf8')
  }
}

class BinaryFile {
  hash: string
  length: number
  preview: string
  constructor(buff: Buffer, trim = 32) {
    this.hash = createHash('sha1').update(buff).digest('hex')
    this.length = buff.length
    const head = buff.subarray(0, trim).toString('base64')
    const tail = buff.subarray(buff.length - trim).toString('base64')
    this.preview = `${head}...${tail}`
  }
}

class Symlink {
  constructor(public target: string) {}
}

class Directory {}
