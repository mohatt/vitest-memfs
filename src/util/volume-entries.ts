import path from 'node:path'
import isGlob from 'is-glob'
import { toRegex } from 'glob-to-regex.js'
import { isPlainObject } from './common.js'
import type { VolumePathType, VolumePathEntry } from './volume.js'

export type VolumeEntryType = Exclude<VolumePathType, 'other'> | 'any'

export interface VolumeEntryRecord {
  type?: VolumeEntryType
  count?: number
}

export interface VolumeEntryArrayRecord extends VolumeEntryRecord {
  path: string
}

export type VolumeEntries =
  | string
  | Array<string | VolumeEntryArrayRecord>
  | Record<string, VolumeEntryType | VolumeEntryRecord>

export type MatchRules = Map<string, MatchRule>

export function createMatchRules(input: VolumeEntries, prefix?: string): MatchRules {
  const basePrefix = resolvePrefix(prefix)
  const rules: MatchRules = new Map()

  function addRule(path: string, opts?: VolumeEntryRecord) {
    const entry = createMatchRule(path, basePrefix, opts)
    rules.set(entry.identifier, entry)
  }

  if (typeof input === 'string') {
    addRule(input)
  } else if (Array.isArray(input)) {
    for (const value of input) {
      if (typeof value === 'string') {
        addRule(value)
      } else if (isPlainObject(value) && typeof value.path === 'string') {
        addRule(value.path, value)
      } else {
        throw new TypeError(
          `Expected array item to be string | { path: string }, got \`${JSON.stringify(value)}\``,
        )
      }
    }
  } else if (isPlainObject(input)) {
    for (const [key, value] of Object.entries(input)) {
      if (typeof value === 'string') {
        addRule(key, { type: value })
      } else if (isPlainObject(value)) {
        addRule(key, value)
      } else {
        throw new TypeError(
          `Expected object value for key \`${key}\` to be string | plain object, got \`${JSON.stringify(value)}\``,
        )
      }
    }
  } else {
    throw new TypeError(`Expected string | array | plain object, got \`${JSON.stringify(input)}\``)
  }

  return rules
}

interface ExactMatchRule {
  kind: 'exact'
  identifier: string
  path: string
  expType: VolumeEntryType
}

interface GlobMatchRule {
  kind: 'glob'
  identifier: string
  pattern: string
  regex: RegExp
  expType: VolumeEntryType
  expCount: number
}

type MatchRule = ExactMatchRule | GlobMatchRule

function createMatchRule(rawPath: string, prefix: string, config?: VolumeEntryRecord): MatchRule {
  const { type, count } = config ?? {}
  if (type != null && !VALID_ENTRY_TYPES.includes(type)) {
    throw new TypeError(
      `Expected entry type to be \`${VALID_ENTRY_TYPES.join('`, `')}\`, got \`${type}\``,
    )
  }

  if (count != null && count <= 0) {
    throw new TypeError(`Expected entry count to be a positive integer, got \`${count}\``)
  }

  const resolved = rawPath.startsWith('/')
    ? path.posix.normalize(rawPath)
    : path.posix.join(prefix, rawPath)
  const expType = type ?? 'any'

  if (isGlobLike(rawPath)) {
    return {
      kind: 'glob',
      identifier: `glob(\`${resolved}\`)`,
      pattern: resolved,
      regex: toRegex(resolved),
      expType,
      expCount: count ?? 1,
    }
  }

  return {
    kind: 'exact',
    identifier: resolved,
    path: resolved,
    expType,
  }
}

interface DiffEntry {
  exists?: boolean
  type?: string
  count?: number
}

export function matchEntries(pathEntries: VolumePathEntry[], rules: MatchRules) {
  const actualDiff: Record<string, DiffEntry> = {}
  const expectedDiff: Record<string, DiffEntry> = {}
  const allMatches: VolumePathEntry[] = []
  let missingCount = 0
  let typeCount = 0

  rules.forEach((rule) => {
    if (rule.kind === 'exact') {
      const match = pathEntries.find(([path]) => path === rule.path)
      if (!match) {
        actualDiff[rule.identifier] = { exists: false }
        expectedDiff[rule.identifier] = { exists: true }
        missingCount++
        return
      }

      if (rule.expType !== 'any' && match[1] !== rule.expType) {
        actualDiff[rule.identifier] = { type: match[1] }
        expectedDiff[rule.identifier] = { type: rule.expType }
        typeCount++
        return
      }

      allMatches.push(match)
      return
    }

    const matches = pathEntries.filter(([path]) => rule.regex.test(path))
    if (matches.length < rule.expCount) {
      actualDiff[rule.identifier] = { count: matches.length }
      expectedDiff[rule.identifier] = { count: rule.expCount }
      missingCount++
      return
    }

    if (rule.expType !== 'any') {
      const typeMatches = matches.filter(([, type]) => type === rule.expType)
      if (typeMatches.length < rule.expCount) {
        actualDiff[rule.identifier] = { count: typeMatches.length }
        expectedDiff[rule.identifier] = { count: rule.expCount }
        typeCount++
        return
      }

      allMatches.push(...typeMatches)
      return
    }

    allMatches.push(...matches)
  })

  return {
    missingCount,
    typeCount,
    diff: { actual: actualDiff, expected: expectedDiff },
    matches: allMatches,
  }
}

function resolvePrefix(prefix?: string) {
  if (!prefix) return '/'
  const resolved = path.posix.resolve('/', prefix)
  return resolved === '/' ? '/' : resolved.replace(/\/$/, '')
}

export function isGlobLike(value: string) {
  return isGlob(value, { strict: false })
}

export function parseRegExp(value: string) {
  const match = /^\/(.{1,4096})\/([gimsuy]{0,6})$/.exec(value)
  if (match) {
    const [, expr, flags] = match
    return new RegExp(expr, flags)
  }
  return null
}

const VALID_ENTRY_TYPES: readonly VolumeEntryType[] = ['file', 'dir', 'symlink', 'any']
