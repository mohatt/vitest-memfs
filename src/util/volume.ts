import path from 'node:path'
import pLimit from 'p-limit'
import type { Volume } from 'memfs'
import { importActualFS } from './common.js'

export type VolumeEntry =
  | { kind: 'file'; data: Buffer }
  | { kind: 'symlink'; target: string }
  | { kind: 'empty-dir' }

export interface VolumeMap {
  [path: string]: VolumeEntry
}

interface VolumeToMapOptions {
  prefix?: string
  withData?: boolean
}

/**
 * Get a filename -> Buffer map from current volume.
 */
export function volumeToMap(volume: Volume, options?: VolumeToMapOptions) {
  const { prefix = '/', withData = true } = options ?? {}
  const map: VolumeMap = Object.create(null)

  function walk(curr: string) {
    const stats = volume.lstatSync(curr)
    if (stats.isDirectory()) {
      const list = volume.readdirSync(curr) as string[]
      if (list.length === 0) map[curr] = { kind: 'empty-dir' }
      for (const name of list) {
        walk(path.posix.join(curr, name))
      }
    } else if (stats.isFile()) {
      map[curr] = {
        kind: 'file', //
        data: withData ? (volume.readFileSync(curr) as Buffer) : Buffer.alloc(0),
      }
    } else if (stats.isSymbolicLink()) {
      map[curr] = {
        kind: 'symlink', //
        target: volume.readlinkSync(curr) as string,
      }
    }
  }

  walk(prefix)
  return map
}

export interface ReadDirToMapOptions extends VolumeToMapOptions {
  concurrency?: number
}

export async function readDirToMap(targetDirPath: string, options?: ReadDirToMapOptions) {
  const fsp = await importActualFS()
  const { prefix = '', withData = true, concurrency = 48 } = options ?? {}
  const map: VolumeMap = Object.create(null)

  const limit = pLimit(concurrency)
  const EMPTY_BUFFER = Buffer.alloc(0)

  async function walk(dirPath: string) {
    const entries = await fsp.readdir(dirPath, { withFileTypes: true })
    if (entries.length === 0) {
      const rel = path.posix.relative(targetDirPath, dirPath)
      map[path.posix.join('/', prefix, rel)] = { kind: 'empty-dir' }
    }

    await Promise.all(
      entries.map(async (entry) => {
        const abs = path.join(dirPath, entry.name)
        const rel = path.posix.relative(targetDirPath, abs)
        const key = path.posix.join('/', prefix, rel)

        if (entry.isDirectory()) {
          await walk(abs)
        } else if (entry.isFile()) {
          map[key] = {
            kind: 'file',
            data: withData ? await limit(() => fsp.readFile(abs)) : EMPTY_BUFFER,
          }
        } else if (entry.isSymbolicLink()) {
          map[key] = {
            kind: 'symlink',
            target: await limit(() => fsp.readlink(abs)),
          }
        }
      }),
    )
  }

  await walk(targetDirPath)
  return map
}

export interface WriteVolumeToDirOptions extends VolumeToMapOptions {
  clear?: boolean
  concurrency?: number
}

export async function writeVolumeToDir(
  volume: Volume,
  targetDirPath: string,
  options?: WriteVolumeToDirOptions,
) {
  const fsp = await importActualFS()
  const { prefix, clear, withData = true, concurrency = 48 } = options ?? {}
  const realPrefix = (prefix ? path.posix.resolve('/', prefix) : '') + '/'
  const map = volumeToMap(volume, { prefix: realPrefix })

  if (clear) {
    await fsp.rm(targetDirPath, { recursive: true, force: true })
  }

  const writeDirs = new Set<string>()
  const writeOps: Array<() => Promise<void>> = []

  for (const abs in map) {
    // strip prefix
    const rel = abs.slice(realPrefix.length)
    const targetPath = path.join(targetDirPath, rel)
    const entry = map[abs]

    if (entry.kind === 'file') {
      writeDirs.add(path.dirname(targetPath))
      writeOps.push(() => fsp.writeFile(targetPath, withData ? entry.data : Buffer.alloc(0)))
    } else if (entry.kind === 'symlink') {
      writeDirs.add(path.dirname(targetPath))
      writeOps.push(async () => fsp.symlink(entry.target, targetPath))
    } else if (entry.kind === 'empty-dir') {
      writeDirs.add(targetPath)
    }
  }

  // ensure directories exist
  await Promise.all(Array.from(writeDirs).map((dir) => fsp.mkdir(dir, { recursive: true })))

  // run file/symlink writes with concurrency limit
  const limit = pLimit(concurrency)
  await Promise.all(writeOps.map((op) => limit(op)))
}
