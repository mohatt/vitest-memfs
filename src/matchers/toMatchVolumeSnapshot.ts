import path from 'path'
import { Volume } from 'memfs'
import { createMatcher, getActualFS } from '@/util/common.js'
import { readDirToMap, volumeToMap, writeVolumeToDir } from '@/util/volume.js'
import { compareVolumeMaps, VolumeCompareOptions } from '@/util/volume-compare.js'

export interface VolumeSnapshotMatcherOptions extends VolumeCompareOptions {
  prefix?: string
}

declare module 'vitest' {
  interface Matchers<T = any> {
    /**
     * Assert that a memfs volume matches the snapshot directory.
     */
    toMatchVolumeSnapshot(snapshotDir: string, options?: VolumeSnapshotMatcherOptions): Promise<T>
  }
}

export default createMatcher(
  'toMatchVolumeSnapshot',
  async function (received, snapshotDir, options) {
    if (this.isNot) {
      throw new Error(`toMatchVolumeSnapshot() cannot be used with "not"`)
    }

    const { currentTestName, snapshotState, utils } = this
    const snapshotDirPath = path.join(path.dirname(snapshotState.snapshotPath), snapshotDir)
    const matchSnapshot = (passed: boolean) => {
      snapshotState.match({
        testId: currentTestName,
        testName: currentTestName,
        received: passed ? 'match' : 'mismatch',
        isInline: false,
        rawSnapshot: {
          file: snapshotDirPath,
          content: 'match',
          readonly: true,
        },
      })
      return passed
    }

    if (!(received instanceof Volume)) {
      return {
        pass: matchSnapshot(false),
        message: () => `expected ${utils.printReceived(received)} to be a memfs Volume instance`,
        actual: received,
        expected: new (class Volume {})(),
      }
    }

    const fs = await getActualFS()
    const updateSnapshot =
      (snapshotState as any)._updateSnapshot === 'all' ||
      (await fs
        .access(snapshotDirPath)
        .then(() => false)
        .catch(() => true))

    const prefix = options?.prefix ?? undefined
    if (updateSnapshot) {
      await writeVolumeToDir(received, snapshotDirPath, { prefix, clear: true })
      return { pass: true, message: () => `updated snapshot at ${snapshotDir}` }
    }

    const expectedMap = await readDirToMap(snapshotDirPath, prefix)
    const receivedMap = volumeToMap(received, prefix)

    const result = compareVolumeMaps(receivedMap, expectedMap, options)
    matchSnapshot(result.pass)
    return result
  },
)
