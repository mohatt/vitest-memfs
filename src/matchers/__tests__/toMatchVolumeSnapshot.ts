import { describe, it, expect, vi } from 'vitest'
import path from 'path'
import fsx from 'fs-extra'
import { makeTests, makeVol, pathToMap, VolumeInput } from '@test/util.js'
import toMatchVolumeSnapshot, { VolumeSnapshotMatcherOptions } from '../toMatchVolumeSnapshot.js'

interface TestCase {
  name: string
  left: VolumeInput
  right?: string
  opts?: VolumeSnapshotMatcherOptions
  update?: 'all' | 'new'
  pass: boolean
}

const newCases = makeTests<TestCase>([
  {
    name: 'empty volume',
    left: {},
    pass: true,
  },
  {
    name: 'empty dir',
    left: { '/empty': null },
    pass: true,
  },
  {
    name: 'invalid volume',
    left: () => 'invalid' as any,
    pass: false,
  },
  {
    name: 'different files',
    left: {
      '/src/index.js': '// hi',
      '/bin.dat': Buffer.from([0xde, 0xad, 0xbe, 0xef]),
      '/.gitignore': '',
    },
    pass: true,
  },
  {
    name: 'prefix option',
    left: { '/src/foo.txt': 'hi', '/bar.txt': 'ignore-me' },
    opts: { prefix: '/src' },
    pass: true,
  },
])

const existingCases = makeTests<TestCase>([
  {
    name: 'empty volume',
    left: {},
    right: 'empty-vol',
    pass: true,
  },
  {
    name: 'empty dir vs missing dir',
    left: { '/empty': null },
    right: 'empty-vol',
    pass: false,
  },
  {
    name: 'empty dir match',
    left: { '/empty': null },
    right: 'empty-dir',
    pass: true,
  },
  {
    name: 'force update',
    left: { '/foo.txt': 'hi' },
    right: 'empty-dir',
    update: 'all',
    pass: true,
  },
  {
    name: 'prefix option',
    left: { '/src/foo.txt': 'hi', '/bar.txt': 'ignore-me' },
    right: 'foo-dir',
    opts: { prefix: '/src' },
    pass: true,
  },
  {
    name: 'prefix option mismatch',
    left: { '/src/foo.txt': 'hiz', '/bar.txt': 'ignore-me' },
    right: 'foo-dir',
    opts: { prefix: '/src' },
    pass: false,
  },
  {
    name: 'ignores extra files',
    left: { '/foo.txt': 'hi', '/bar.txt': 'hey', '/extra.txt': 'extra' },
    right: 'foo-bar',
    pass: true,
    opts: { listMatch: 'ignore-extra' },
  },
  {
    name: 'ignores extra files (mismatch)',
    left: { '/foo.txt': 'hi', '/extra.txt': 'extra' },
    right: 'foo-bar',
    pass: false,
    opts: { listMatch: 'ignore-extra' },
  },
  {
    name: 'ignores missing files',
    left: { '/foo.txt': 'hi' },
    right: 'foo-bar',
    pass: true,
    opts: { listMatch: 'ignore-missing' },
  },
  {
    name: 'ignores missing files (mismatch)',
    left: { '/foo.txt': 'hi', '/extra.txt': 'extra' },
    right: 'foo-bar',
    pass: false,
    opts: { listMatch: 'ignore-missing' },
  },
  {
    name: 'binary mismatch',
    left: { '/data.bin': Buffer.alloc(100_000, 0xbb) },
    right: 'bin-dir',
    pass: false,
  },
])

describe('toMatchVolumeSnapshot()', () => {
  describe('unit', () => {
    const mockState = (updateSnapshot: string) => ({
      currentTestName: expect.getState().currentTestName,
      snapshotState: {
        _updateSnapshot: updateSnapshot,
        snapshotPath: path.join(__dirname, '__snapshots__', 'temp', 'xxx.snap'),
        match: vi.fn(),
      },
      utils: {
        printReceived: (received) => `received(${JSON.stringify(received)})`,
        matcherHint: (matcherName: string) => `hint(${matcherName})`,
      },
    })

    async function runTest({ name, left, right, opts, update }: TestCase, existing = false) {
      const leftVol = makeVol(left)
      const rightVal = right ?? name.toLowerCase().replace(/\W+/g, '-')
      const state = mockState(update ?? 'new')
      const invoke = async () => {
        try {
          const matcher = toMatchVolumeSnapshot.bind(state as any)
          const result = await matcher(leftVol, rightVal, opts)
          return { ...result, message: result.message() }
        } catch (e) {
          return e
        }
      }
      const snapDir = path.join(path.dirname(state.snapshotState.snapshotPath), rightVal)
      await fsx.remove(snapDir)
      if (existing) {
        const fixtureDir = path.join(__dirname, '__fixtures__', rightVal)
        await fsx.copy(fixtureDir, snapDir)
      }
      const result = await invoke()
      expect(result).toMatchSnapshot('result')
      if (!existing || update === 'all') {
        const snapDirMap = await pathToMap(snapDir).catch((e) => `${e.name}: ${e.code}`)
        expect(snapDirMap).toMatchSnapshot('disk-snapshot')
      }
      await fsx.remove(snapDir)
    }

    async function testRunnerNew(testCase: TestCase) {
      await runTest(testCase, false)
    }

    async function testRunnerExisting(testCase: TestCase) {
      await runTest(testCase, true)
    }

    it.each(newCases.normal)('$name [write]', testRunnerNew)
    it.only.each(newCases.only)('$name [write]', testRunnerNew)
    it.skip.each(newCases.skip)('$name [write]', testRunnerNew)

    it.each(existingCases.normal)('$name [existing]', testRunnerExisting)
    it.only.each(existingCases.only)('$name [existing]', testRunnerExisting)
    it.skip.each(existingCases.skip)('$name [existing]', testRunnerExisting)
  })

  describe('integration', () => {
    it('works correctly', async () => {
      const vol = makeVol({
        '/foo.txt': 'hi',
        '/bin/data.bin': Buffer.alloc(100_000, 0xbb),
      })
      await expect(vol).toMatchVolumeSnapshot('test')
    })

    it('throws when used with not', async () => {
      const vol = makeVol({ '/foo.txt': 'hi' })
      await expect(
        () => expect(vol).not.toMatchVolumeSnapshot('test'), //
      ).rejects.toThrow(/cannot be used with "not"/)
    })
  })
})
