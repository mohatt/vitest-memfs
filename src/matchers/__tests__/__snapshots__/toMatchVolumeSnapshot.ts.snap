// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toMatchVolumeSnapshot() > unit > 'binary mismatch' [existing] > result 1`] = `
{
  "actual": Buffer {
    "length": 100000,
    "preview": "u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7s=...u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7u7s=",
    "sha1": "84a8811d7cd58ccbe862c1f94a00366c12440816",
  },
  "expected": Buffer {
    "length": 100000,
    "preview": "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=...qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqo=",
    "sha1": "cd561c0d1e53dbd0cd390f4b2cd3728db7dd71fc",
  },
  "message": "binary mismatch in file \`/data.bin\`",
  "pass": false,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'different files' [write] > disk-snapshot 1`] = `
{
  "/.gitignore": [
    "file",
    "",
  ],
  "/bin.dat": [
    "file",
    "3q2+7w==",
  ],
  "/src/index.js": [
    "file",
    "Ly8gaGk=",
  ],
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'different files' [write] > result 1`] = `
{
  "message": "updated snapshot at different-files",
  "pass": true,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'empty dir match' [existing] > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'empty dir vs missing dir' [existing] > result 1`] = `
{
  "actual": [
    "/empty",
  ],
  "expected": [
    "/",
  ],
  "message": "directory structure didnâ€™t match",
  "pass": false,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'empty dir' [write] > disk-snapshot 1`] = `
{
  "/empty": null,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'empty dir' [write] > result 1`] = `
{
  "message": "updated snapshot at empty-dir",
  "pass": true,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'empty volume' [existing] > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'empty volume' [write] > disk-snapshot 1`] = `
{
  "/": null,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'empty volume' [write] > result 1`] = `
{
  "message": "updated snapshot at empty-volume",
  "pass": true,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'force update' [existing] > disk-snapshot 1`] = `
{
  "/foo.txt": [
    "file",
    "aGk=",
  ],
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'force update' [existing] > result 1`] = `
{
  "message": "updated snapshot at empty-dir",
  "pass": true,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'ignores extra files (mismatch)' [existing] > result 1`] = `
{
  "actual": [
    "/foo.txt",
  ],
  "expected": [
    "/bar.txt",
    "/foo.txt",
  ],
  "message": "volume is missing 1 expected file(s)",
  "pass": false,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'ignores extra files' [existing] > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'ignores missing files (mismatch)' [existing] > result 1`] = `
{
  "actual": [
    "/extra.txt",
    "/foo.txt",
  ],
  "expected": [
    "/foo.txt",
  ],
  "message": "volume has 1 unexpected file(s)",
  "pass": false,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'ignores missing files' [existing] > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'invalid volume' [write] > disk-snapshot 1`] = `"Error: ENOENT"`;

exports[`toMatchVolumeSnapshot() > unit > 'invalid volume' [write] > result 1`] = `
{
  "actual": "invalid",
  "expected": Volume {},
  "message": "expected received("invalid") to be a memfs Volume instance",
  "pass": false,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'prefix option mismatch' [existing] > result 1`] = `
{
  "actual": "hiz",
  "expected": "hi",
  "message": "mismatch in file \`/src/foo.txt\`",
  "pass": false,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'prefix option' [existing] > result 1`] = `
{
  "message": "volumes matched",
  "pass": true,
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'prefix option' [write] > disk-snapshot 1`] = `
{
  "/foo.txt": [
    "file",
    "aGk=",
  ],
}
`;

exports[`toMatchVolumeSnapshot() > unit > 'prefix option' [write] > result 1`] = `
{
  "message": "updated snapshot at prefix-option",
  "pass": true,
}
`;
